### PR TITLE
Report generation tool supports GCS path for input/ouput paths

### PR DESCRIPTION
**Pull request by erfanio**
Using `gs://` paths the report generation tool directly read/write files directly from/to GCS.

The experiment nodes will continue using local file system. But now we have additional flexibility to regenerate reports without downloading the results which can be very big (hundreds of GBs for "all" benchmark set). This is a necessary first step for implementing trends report (go/oss-fuzz-gen-reports) which will require back-filling existing reports.

Testing it on a daily comparison result, it took ~11 minutes to generate. This is still pretty slow but we can optimize further later. For now this is good enough if we need to regenerate reports after an experiment is done.

### CODE REVIEW

**Comment by DonggeLiu** [comment]
> gcbrun exp -n dg

Just a side note: Please feel free to use other names for `-n`.
I use `dg` only because that's my name (DongGe)  : )

File: report/web.py
Comment (DonggeLiu) [inline report/web.py:96]: This already looks good to me, but some nits:
1. Instead of checking `if self._gcs_bucket is not None` in every function to know if this is the local filesystem or GCS bucket, would it be a good idea to use class inheritance and overwrite? E.g., `class LocalResultDir` and `class GCSResultDir(LocalResultDir)`.
2. How about hosting this class in its own file?

File: report/web.py
Comment (DonggeLiu) [inline report/web.py:200]: nit: `blob.size if blob.size else 0` or even `blob.size or 0`

**Review by DonggeLiu — APPROVED** [review]
Thanks, @erfanio !
This makes the code a lot more organized.
Approving with some nit discussions.

**Comment by erfanio** [comment]
oops completely missed these comment before merging... :sweat:

I'll have another PR to address these.

File: report/web.py
Comment (erfanio) [reply report/web.py:113]: IMO this implicit false can be a footgun when None is involved. The internal python style also warns against it go/pystyle#truefalse-evaluations-decision

**Comment by DonggeLiu** [comment]
LOL, thanks.

File: report/web.py
Comment (erfanio) [reply report/web.py:130]: Same as above. go/pystyle#truefalse-evaluations-decision

File: report/web.py
Comment (erfanio) [reply report/web.py:155]: Hmmm I'm not sure if `os.path.isfile() or os.path.isdir() == os.path.exists()` is guaranteed to be true in all file systems.

In GCS this holds because it's not a real filesystem.

File: report/web.py
Comment (erfanio) [reply report/web.py:167]: I'll include this fix in the next pull request :)

File: report/web.py
Comment (erfanio) [reply report/web.py:97]: I initially did try to have them be separate classes but to make that work cleanly, I would have to have a abstract base class, a GCS class, a local disk class, and a factory function that creates the right object based on the path. It just ended up being a lot of fluff code.

File: report/web.py
Comment (erfanio) [reply report/web.py:178]: I think keeping this consistent makes it easier for the readers. The early return isn't great but switching up the condition is more confusing IMO.

File: report/web.py
Comment (erfanio) [reply report/web.py:201]: Same as above. go/pystyle#truefalse-evaluations-decision it's better to have None comparisons explicit.

File: report/web.py
Comment (erfanio) [reply report/web.py:167]: Actually it needs to be either, `bool(self.listdir())` or `len(self.listdir()) > 0` to make sure we're actually returning a bool. I think I still prefer the latter.

File: report/web.py
Comment (DonggeLiu) [reply report/web.py:113]: It's a nit, so I don't mind leaving this, just to explain my thought:
I reckon that guideline does not apply here, given we already know `_gcs_client` won't be other types.
It makes more sense for cases where the type is unclear, e.g., searching for an element in a list, which can return `0` (1st element) or `None` (unfound).

File: report/web.py
Comment (DonggeLiu) [reply report/web.py:155]: I think that's a valid assumption, but don't mind leaving it for now as you have merged.

File: report/web.py
Comment (DonggeLiu) [reply report/web.py:97]: > a abstract base class, a GCS class, a local disk class,

I reckon two classes should suffice for now: a base `class LocalResultDir` and an extended `class GCSResultDir(LocalResultDir)`.

> a factory function that creates the right object based on the path.

Would this be a workable solution for now?
```python
if path.startswith('gs://'):
  GCSResultDir(self._results_dir)
else:
  LocalResultDir(self._results_dir)
```

Again, the original implementation LGTM, no need to worry too much about this.

Comment (erfanio) [reply report/web.py:97]: Ah yeah I didn't consider using the local filesystem as a base class. I think both of these designs have their own drawbacks. Happy to chat about this next time we meet.

Same with a lot of these comments, I think we both have different styles that are both valid :)

**Pull request merged**
